### PR TITLE
[Preview] Integrated cache: Adds BypassIntegratedCache to DedicatedGatewayRequestOptions

### PR DESCRIPTION
**Pull request by jiajunpeng-msft**
# Pull Request Template

## Description

This is to add a new DedicatedGatewayRequestOption "BypassIntegratedCache".

Currently, integrated cache is used by default for Dedicated Gateway. Customers cannot skip cache for particular requests or data unless they shift to multi-tenant Gateway,which will lose the benefits of Dedicated Gateway.

For customers to have more control over integrated cache, we're introducing a new "RequestOption" called "BypassIntegratedCache". This option will allow the customer to decide whether to use integrated cache for each request or not. If this value is set to true, the item/query will be served from backend and won't be cached in Dedicated Gateway.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

**Comment by ealsur** [comment]
@msftbot wait for @kirankumarkolli to approve, this PR modifies public API surface

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 1]
```diff
@@ -23,6 +23,12 @@ public class DedicatedGatewayRequestOptions
         /// </remarks> 
         public TimeSpan? MaxIntegratedCacheStaleness { get; set; }
 
+        /// <summary>
+        /// Gets or sets if bypass the integrated cache or not associated with the request in the Azure CosmosDB service.
+        /// </summary>
+        /// <value>Default value is false.</value>
+        public bool? BypassIntegratedCache { get; set; }
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:30]: If this is for Public Preview, it should not be public, it should be conditionally public. For example: 
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/17203a5966e906e0d66d52a95cf4d482add3c51e/Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs#L1018-L1023

Or is this already GAed?

Is this already planned to be added to other SDKs?

**Review by ealsur — CHANGES_REQUESTED** [review]
Waiting on confirmation wheter this is preview or GA

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:30]: Which SDK is it on is orthogonal. Is this feature GA? Or are you introducing it for users to test it (public preview)?

Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:30]: After discussed with my manager, we decided to push this under public preview. Will update the PR soon. Previously we were thinking to directly GA this feature as this is a small change but feels like under public preview is safer.

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 2]
```diff
@@ -36,6 +47,11 @@ internal static void PopulateMaxIntegratedCacheStalenessOption(DedicatedGatewayR
 
                 request.Headers.Set(HttpConstants.HttpHeaders.DedicatedGatewayPerRequestCacheStaleness, cacheStalenessInMilliseconds.ToString(CultureInfo.InvariantCulture));
             }
+
+            if (dedicatedGatewayRequestOptions?.BypassIntegratedCache != null)
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:51]: This is a `bool?` the correct way would be to use HasValue to check if it has been set, and Value to extract the value and compare if it's true.

Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:51]: Please add tests to confirm the header is set only when the property is set to True, and not when it's not set or set to False

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs [hunk 3]
```diff
@@ -23,6 +23,17 @@ public class DedicatedGatewayRequestOptions
         /// </remarks> 
         public TimeSpan? MaxIntegratedCacheStaleness { get; set; }
 
+        /// <summary>
+        /// Gets or sets if bypass the integrated cache or not associated with the request in the Azure CosmosDB service.
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:27]: Please expand with more context on what it does if possible, with an example. 
At-least we need to cover that it doesn't server from cache and also the response is also not cached.

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 2 above)
Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs]: Updated.

File: Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs (hunk 3 above)
Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos/src/Resource/Settings/DedicatedGatewayRequestOptions.cs:27]: Updated.

File: Microsoft.Azure.Cosmos.Encryption.Custom/tests/Microsoft.Azure.Cosmos.Encryption.Custom.Tests/Contracts/DotNetSDKEncryptionCustomAPI.json [hunk 4]
```diff
@@ -51,6 +51,13 @@
           ],
           "MethodInfo": "System.Threading.Tasks.Task`1[Microsoft.Azure.Cosmos.Encryption.Custom.DataEncryptionKey] FetchDataEncryptionKeyAsync(System.String, System.String, System.Threading.CancellationToken);IsAbstract:False;IsStatic:False;IsVirtual:True;IsGenericMethod:False;IsConstructor:False;IsFinal:False;"
         },
+        "System.Threading.Tasks.Task`1[Microsoft.Azure.Cosmos.Encryption.Custom.DataEncryptionKey] FetchDataEncryptionKeyWithoutRawKeyAsync(System.String, System.String, System.Threading.CancellationToken)[System.Runtime.CompilerServices.AsyncStateMachineAttribute(typeof(Microsoft.Azure.Cosmos.Encryption.Custom.CosmosDataEncryptionKeyProvider+<FetchDataEncryptionKeyWithoutRawKeyAsync>))]": {
+          "Type": "Method",
+          "Attributes": [
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos.Encryption.Custom/tests/Microsoft.Azure.Cosmos.Encryption.Custom.Tests/Contracts/DotNetSDKEncryptionCustomAPI.json:56]: Does this PR includes change n encryption project also? Please revert this file if not.

Comment (jiajunpeng-msft) [reply Microsoft.Azure.Cosmos.Encryption.Custom/tests/Microsoft.Azure.Cosmos.Encryption.Custom.Tests/Contracts/DotNetSDKEncryptionCustomAPI.json]: Reverted.

**Review by ealsur — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Pull request merged**